src/transations.rs

```diff
@@ -0,0 +1,1359 @@
+                    sighash(&tx, input_index, &txo, &script_code, is_anyonecanpay)
+                ),
+                _ => Err(
+                    RevaultError::Signature(
+                        "Wrong transaction output type: cancel transactions only spend unvault transactions and fee-bumping transactions"
+                        .to_string()
+                    )
+                ),
+            },
+            RevaultTransaction::EmergencyTransaction(ref tx) => match previous_txout {
+                RevaultTxOut::VaultTxOut(ref txo)
+                | RevaultTxOut::UnvaultTxOut(ref txo)
+                | RevaultTxOut::FeeBumpTxOut(ref txo) => Ok(
+                    sighash(&tx, input_index, &txo, &script_code, is_anyonecanpay)
+                ),
+                _ => Err(
+                    RevaultError::Signature(
+                        "Wrong transaction output type: emergency transactions only spend vault, unvault and fee-bumping transactions"
+                        .to_string()
+                    )
+                ),
+            }
+        }
+    }
+
+    /// Verify this transaction validity against libbitcoinconsensus.
+    /// Handles all the destructuring and txout research internally.
+    ///
+    /// # Errors
+    /// - If verification fails.
+    pub fn verify(
+        &self,
+        previous_transactions: &[&RevaultTransaction],
+    ) -> Result<(), RevaultError> {
+        // Look for a referenced txout in the set of spent transactions
+        // TODO: optimize this by walking the previous tx set only once ?
+        fn get_txout(prevout: &OutPoint, transactions: &[&RevaultTransaction]) -> Option<TxOut> {
+            for prev_tx in transactions {
+                match *prev_tx {
+                    RevaultTransaction::VaultTransaction(ref tx)
+                    | RevaultTransaction::UnvaultTransaction(ref tx)
+                    | RevaultTransaction::SpendTransaction(ref tx)
+                    | RevaultTransaction::CancelTransaction(ref tx)
+                    | RevaultTransaction::EmergencyTransaction(ref tx)
+                    | RevaultTransaction::FeeBumpTransaction(ref tx) => {
+                        if tx.txid() == prevout.txid {
+                            if prevout.vout as usize >= tx.output.len() {
+                                return None;
+                            }
+                            return Some(tx.output[prevout.vout as usize].clone());
+                        }
+                    }
+                }
+            }
+
+            None
+        }
+
+        match *self {
+            RevaultTransaction::VaultTransaction(ref tx)
+            | RevaultTransaction::UnvaultTransaction(ref tx)
+            | RevaultTransaction::SpendTransaction(ref tx)
+            | RevaultTransaction::CancelTransaction(ref tx)
+            | RevaultTransaction::EmergencyTransaction(ref tx)
+            | RevaultTransaction::FeeBumpTransaction(ref tx) => {
```

Comment:
what do you think of implementing a method for VaultTransaction to get the tx ? I think the pattern matching will be often repetitive

Reply:
Yeah, it is (but at least only on the inside). I was thinking of a macro but a method is OK too